src/target/riscv/riscv-013.c

```diff
@@ -2073,3 +2076,418 @@ void riscv013_clear_abstract_error(struct target *target)
 	// Clear the error status.
 	dmi_write(target, DMI_ABSTRACTCS, abstractcs & DMI_ABSTRACTCS_CMDERR);
 }
+
+#define COMPLIANCE_TEST(b, message) {		\
+    int pass = 0;				\
+    if (b) {					\
+      pass = 1;					\
+      passed_tests ++;				\
+    }									\
+    LOG_INFO("%s test %d (%s)\n", (pass) ? "PASSED":"FAILED",  total_tests, message); \
+    assert(pass);							\
+    total_tests ++;							\
+  }
+   
+int riscv013_test_compliance(struct target *target) {
+  LOG_INFO("Testing Compliance against RISC-V Debug Spec v0.13");
+
+  if (!riscv_rtos_enabled(target)) {
+    LOG_ERROR("Please run with -rtos riscv to run compliance test.");
+    return ERROR_FAIL;
+  }
+
+  int total_tests = 0;
+  int passed_tests = 0;
+
+  uint32_t dmcontrol_orig = dmi_read(target, DMI_DMCONTROL);
+  uint32_t dmcontrol;
+  uint32_t testvar;
+  dmcontrol = set_field(dmcontrol_orig, DMI_DMCONTROL_HARTSEL, RISCV_MAX_HARTS-1);
+  dmi_write(target, DMI_DMCONTROL, dmcontrol);
+  dmcontrol = dmi_read(target, DMI_DMCONTROL);
+  COMPLIANCE_TEST(get_field(dmcontrol, DMI_DMCONTROL_HARTSEL) == (RISCV_MAX_HARTS-1), "DMCONTROL.hartsel should hold MAX_HARTS - 1");
```

Comment:
This needs to be updated based on spec now allowing these bit to not be fully RW